cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md

```diff
@@ -91,33 +91,27 @@ For these cases, attempting to retrofit this pattern in existing applications re
 
 ### Encryption based Token Pattern
 
-The Encrypted Token Pattern leverages an encryption, rather than comparison method of Token-validation. It is most suitable for applications that do not want to maintain any state at server side.
+The Encrypted Token Pattern leverages an encryption, rather than comparison method of Token-validation. It is most suitable for applications that do not want to maintain any state at server side. 
 
-After successful authentication, the server generates a unique token comprised of the user's ID, a timestamp value and a [nonce](http://en.wikipedia.org/wiki/Cryptographic_nonce), using a unique key available only on the server. This token is returned to the client and embedded in a hidden field for forms, in the request-header/parameter for AJAX requests. On receipt of this request, the server reads and decrypts the token value with the same key used to create the token. Inability to correctly decrypt suggest an intrusion attempt. Once decrypted, the UserId and timestamp contained within the token are validated; the UserId is compared against the currently logged in user, and the timestamp is compared against the current time.
+Server generates a token comprised of the user's session ID and a timestamp (to prevent replay attacks) using a unique key available only on the server (AES256 or stronger encryption algorithms are recommended). This token is returned to the client and embedded in a hidden field for forms, in the request-header/parameter for AJAX requests. On receipt of this request, the server reads and decrypts the token value with the same key used to create the token. Inability to correctly decrypt suggest an intrusion attempt (recommend to block and log the attack for incident response purposes). Once decrypted, the users sessionId and timestamp contained within the token are validated; session-id is compared against the currently logged in user, and the timestamp is compared against the current time to verify that its not beyond the defined token expiry time. If session-id matches and the timestamp is under the defined token expiry time, request can be allowed.
```

Comment:
About encryption using AES algorithm, it can be interesting to warn the reader to use a [mode of operation](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) as well as an implementation that do not expose the ciphered token to [padding oracle attack](https://www.owasp.org/index.php/Testing_for_Padding_Oracle_(OTG-CRYPST-002)) and propose the reader to use [AES with GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) or [AES with SIV](https://github.com/miscreant/miscreant/wiki/AES-SIV)  for example.

What do you think?

Reply:
Hey, 

Yes, its a valid point. I thought of keeping it in the beginning but then I thought it might look more of cryptography document and did not keep it. I will keep it now.  

Also, I was trying to see if I can link any OWASP cryptography document and then found [this](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cryptographic_Storage_Cheat_Sheet.md) and [this](https://www.owasp.org/index.php/Testing_for_Weak_Encryption_(OTG-CRYPST-004)) where we still recommend AES 128. Maybe we should update them recommending to 256 bits?

Reply:
*I am not a crypto-analyst so perhaps the thing i said below are totally wrong:*

I have seen that there no signifiant security strengh increase between AES 256 vs 128 bits when the target protected data is not a critical asset. I think here that the important point is the fact that the anti-CSRF token is ciphered but we can accept the risk that it can be theoretically "cracked".

https://www.keylength.com/en/4/

![Keylength-NIST-Reco](https://user-images.githubusercontent.com/1573775/56747256-e7db6980-677d-11e9-8bec-8f4bba48159b.png).

As you said references to the *Cryptographic Storage Cheat Sheet* can be useful and we can, in this section of the CSRF CS, just add a simple note about the importance of using GCM or SIV mode of operation and not CBC based mode of operation.

What do you think?

Reply:
Yes, I agree with what you quoted for 128 but just as a defense in depth/safe side its better to recommend 256. 128 is OK too.

I think CBC is safe to use as long as few conditions are met. ECB is strictly not recommended. Made changes accordingly.

Reply:
If you agree, you can resolve this conversation.

Reply:
My following word has to be confirmed with @ThunderSon and @mackowski and @jmanico

I think that it's preferable to propose to use `AES 128 GCM/SIV` over `AES 256 CBC` because:
* A strengh of 128 bits seems to be enough here according to the target value protected.
* CBC seems to tend to be more and more deprecated in flavor of GCM/SIV: [link](https://security.stackexchange.com/a/150749) and [link](https://security.googleblog.com/2013/11/a-roster-of-tls-cipher-suites-weaknesses.html)

I'm not blocked on the topic, I just try to find to right balance and advice for a developer implementing the crypto on the CSRF token and not aware about CBC vs GCM vs SIV usage context and caveats...

Reply:
I have read your update on this part and it's OK for me.
Thanks for the update.